### PR TITLE
Dcm 125 add retry for messages

### PR DESCRIPTION
**Pull request by dashton82**
- Added retry functionality using Polly.
- Update the authentication to use Azure AD.
- Moved the API client into this solution from shared packages

Once merged I'll update the usages of this across environments for Provider, EAS and Employer Users. When running locally you **no longer need to provide authentication credentials**.

Also need to look at:
- implementing Audit
- Remove table storage usage
- Downgrade to 4.5 from 4.5.2

### CODE REVIEW

**Comment by mgwilliam** [comment]
cheers dan - looks comprehensive! 
sorry, not looked at the code yet - couple of questions from the commits though:

1. do the auth changes require consumers to be updated? if so, we may need to split some of this into separate PRs so we can deploy the notification improvements without having to change consumers at the same time

2. will the retry handle prolonged gov.uk / other outages? ie. not just intermittent / "in proc" failures

can catch up tomorrow if this doesn't make sense :)

**Comment by dashton82** [comment]
@mgwilliam yep - at the moment if its a too many requests response it will retry forever, for 503's it will retry 5 times leaving 10s between. We can perhaps refine these policies/timings and also add any extra exceptions that we've seen. 

The auth change will require a minor change to the consumer - although i can very easily put the old JWT back in and not release the client then everything will continue to work as is.

File: src/SFA.DAS.Notifications.Domain/Http/ServiceUnavailableException.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿namespace SFA.DAS.Notifications.Domain.Http
+{
+    public class ServiceUnavailableException : HttpException
+    {
+        public ServiceUnavailableException()
+            : base(500, "Service is unavailable")
```
Comment (mgwilliam) [inline src/SFA.DAS.Notifications.Domain/Http/ServiceUnavailableException.cs]: should be 503 ?

File: src/SFA.DAS.Notifications.Infrastructure/ExecutionPolicies/SendEmailExecutionPolicy.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            _logger = logger;
+
+            TooManyRequestsPolicy = Policy.Handle<TooManyRequestsException>().WaitAndRetryForeverAsync((i) => new TimeSpan(0, 0, 10), (ex, ts) => OnRetryableFailure(ex));
+            ServiceUnavailablePolicy = CreateAsyncRetryPolicy<ServiceUnavailableException>(5, new TimeSpan(0, 0, 10), OnRetryableFailure);
+            RootPolicy = Policy.WrapAsync(TooManyRequestsPolicy, ServiceUnavailablePolicy);
+        }
+
+        protected override T OnException<T>(Exception ex)
+        {
+            _logger.Error(ex, $"Exceeded retry limit - {ex.Message}");
```
Comment (mgwilliam) [inline src/SFA.DAS.Notifications.Infrastructure/ExecutionPolicies/SendEmailExecutionPolicy.cs]: is this message correct?

Comment (dashton82) [inline src/SFA.DAS.Notifications.Infrastructure/ExecutionPolicies/SendEmailExecutionPolicy.cs]: That one is correct - its the general catch all and is logged once the retry limit for each handled exception has been reached

**Comment by mgwilliam** [comment]
will need to remove the config for jwts too when move to aad

File: src/SFA.DAS.Notifications.Infrastructure/ExecutionPolicies/SendEmailExecutionPolicy.cs (hunk 2 above)
Comment (mgwilliam) [inline src/SFA.DAS.Notifications.Infrastructure/ExecutionPolicies/SendEmailExecutionPolicy.cs]: ok cool

**Pull request merged**